service/history/historyReplicator.go

```diff
@@ -440,14 +375,184 @@ func (r *historyReplicator) FlushBuffer(context *workflowExecutionContext, msBui
 		}
 
 		// Apply replication task to workflow execution
-		if err := r.ApplyReplicationTask(context, msBuilder, req); err != nil {
+		if err := r.ApplyReplicationTask(context, msBuilder, req, logger); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
+func (r *historyReplicator) replicateWorkflowStarted(context *workflowExecutionContext, msBuilder *mutableStateBuilder, di *decisionInfo, history *shared.History,
+	sBuilder *stateBuilder, requestID string, logger bark.Logger) error {
+	domainID := msBuilder.executionInfo.DomainID
+	execution := shared.WorkflowExecution{
+		WorkflowId: common.StringPtr(msBuilder.executionInfo.WorkflowID),
+		RunId:      common.StringPtr(msBuilder.executionInfo.RunID),
+	}
+	var parentExecution *shared.WorkflowExecution
+	initiatedID := common.EmptyEventID
+	parentDomainID := ""
+	if msBuilder.executionInfo.ParentDomainID != "" {
+		initiatedID = msBuilder.executionInfo.InitiatedID
+		parentDomainID = msBuilder.executionInfo.ParentDomainID
+		parentExecution = &shared.WorkflowExecution{
+			WorkflowId: common.StringPtr(msBuilder.executionInfo.ParentWorkflowID),
+			RunId:      common.StringPtr(msBuilder.executionInfo.ParentRunID),
+		}
+	}
+	firstEvent := history.Events[0]
+	lastEvent := history.Events[len(history.Events)-1]
+
+	// Serialize the history
+	serializedHistory, serializedError := r.Serialize(history)
+	if serializedError != nil {
+		logging.LogHistorySerializationErrorEvent(logger, serializedError, fmt.Sprintf(
+			"HistoryEventBatch serialization error on start workflow.  WorkflowID: %v, RunID: %v",
+			execution.GetWorkflowId(), execution.GetRunId()))
+		return serializedError
+	}
+
+	// Generate a transaction ID for appending events to history
+	transactionID, err := r.shard.GetNextTransferTaskID()
+	if err != nil {
+		return err
+	}
+
+	err = r.shard.AppendHistoryEvents(&persistence.AppendHistoryEventsRequest{
+		DomainID:      domainID,
+		Execution:     execution,
+		TransactionID: transactionID,
+		FirstEventID:  firstEvent.GetEventId(),
+		Events:        serializedHistory,
+	})
+	if err != nil {
+		return err
+	}
+
+	// TODO this pile of logic should be merge into workflow execution context / mutable state
+	msBuilder.executionInfo.LastFirstEventID = firstEvent.GetEventId()
+	msBuilder.executionInfo.NextEventID = lastEvent.GetEventId() + 1
+	incomingVersion := firstEvent.GetVersion()
+	replicationState := &persistence.ReplicationState{
```

Comment:
Probably we should get rid of this and just call:
msBuilder.updateReplicationStateLastEventID